### PR TITLE
Feature/9119 booster notification/9218 invalid signature notifications

### PR DESCRIPTION
**Pull request by pascal-brause**
## Description
Adds notifications to be scheduled and triggered when a certificate gets `.invalid`. A Notification is scheduled immediately. When opening the notification, a route is created and leads to the certificate's details screen.
New is now the check for the validity check in the background.

I also added in the dev menu entry for scheduling notifications for a specific certificate the entry for invalid.

The amount of file changed comes by a renaming from `signatureVerifying` to `dccSignatureVerifier`.

## Link to Jira
Mainstory: https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-9119
Subtask: https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-9218

## Screenshots
9218-see notification
9218-devmenu
iOS

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift [hunk 1]
```diff
@@ -1731,13 +1731,53 @@ class HealthCertificateServiceTests: CWATestCase {
...
+			store: store,
+			dccSignatureVerifier: DCCSignatureVerifyingStub(error: .HC_DSC_EXPIRED),
+			dscListProvider: MockDSCListProvider(),
+			client: ClientMock(),
+			appConfiguration: CachedAppConfigurationMock(),
+			digitalCovidCertificateAccess: MockDigitalCovidCertificateAccess(),
+			notificationCenter: notificationCenter
+		)
+		
+		// There should be now 1 notifications for invalid, 1 for expireSoon and 1 for expired.
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift:1771]:
```suggestion
		// There should be now 1 notification for invalid, 1 for expireSoon and 1 for expired.
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift [hunk 2]
```diff
@@ -849,6 +861,13 @@ class HealthCertificateService {
 		let expirationDate = healthCertificate.expirationDate
 		scheduleNotificationForExpiringSoon(id: id, date: expiringSoonDate)
 		scheduleNotificationForExpired(id: id, date: expirationDate)
+
+		// Schedule an 'invalid' notification, if it was not scheduled before.
+		if healthCertificate.validityState == .invalid && !healthCertificate.didShowInvalidNotification {
+
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/HealthCertificateService.swift:867]:
```suggestion
```

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift [hunk 3]
```diff
@@ -107,6 +111,23 @@ final class DMLocalNotificationsViewModel {
 
 		addNotification(request: request)
 	}
+	private func scheduleNotificationForInvalid(
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift:114]:
```suggestion
	private func scheduleNotificationForInvalidCertificate(
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/ENATaskExecutionDelegate.swift [hunk 4]
```diff
@@ -300,4 +314,8 @@ class TaskExecutionHandler: ENATaskExecutionDelegate {
 			completion()
 		})
 	}
+
+	private func checkCertificateValidityStates(completion: @escaping () -> Void) {
+		healthCertificateService.updateValidityStatesAndNotificationsWithFreshDSCList(shouldScheduleTimer: false, completion: completion)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/ENATaskExecutionDelegate.swift:319]:
```suggestion
		healthCertificateService.updateValidityStatesAndNotifiesWithFreshDSCList(shouldScheduleTimer: false, completion: completion)
```

File: src/xcode/ENA/ENA/Source/AppDelegate & Globals/ENATaskExecutionDelegate.swift [hunk 5]
```diff
@@ -18,22 +20,24 @@ class TaskExecutionHandler: ENATaskExecutionDelegate {
...
 		eventCheckoutService: EventCheckoutService,
 		store: Store,
-		exposureSubmissionDependencies: ExposureSubmissionServiceDependencies
+		exposureSubmissionDependencies: ExposureSubmissionServiceDependencies,
+		healthCertificateService: HealthCertificateService
 	) {
 		self.riskProvider = riskProvider
 		self.exposureManager = exposureManager
-		self.pdService = plausibleDeniabilityService
+		self.plausibleDeniabilityService = plausibleDeniabilityService
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/AppDelegate & Globals/ENATaskExecutionDelegate.swift:28]: 👍

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift [hunk 6]
```diff
@@ -107,6 +111,23 @@ final class DMLocalNotificationsViewModel {
...
+		id: String
+	) {
+		let content = UNMutableNotificationContent()
+		content.title = AppStrings.LocalNotifications.certificateGenericTitle
+		content.body = AppStrings.LocalNotifications.certificateGenericBody
+		content.sound = .default
+
+		let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 2.0, repeats: false)
```
Comment (naveeddotio) [inline src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift:122]: Does the tech-specs says to trigger it after two seconds?

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift [hunk 7]
```diff
@@ -33,13 +34,15 @@ final class HealthCertificate: Encodable, Equatable, Comparable {
 	enum CodingKeys: String, CodingKey {
 		case base45
 		case validityState
+		case didShowInavlidNotification
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/Models/HealthCertificate.swift:37]:
```suggestion
		case didShowInvalidNotification
```

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift [hunk 8]
```diff
@@ -1731,13 +1731,53 @@ class HealthCertificateServiceTests: CWATestCase {
...
+		XCTAssertEqual(notificationCenter.notificationRequests.count, 3)
```
Comment (ArturFriesen) [inline src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift:1772]: Is it correct, that 1 certificate can trigger 3 notifications for the different states?

Comment (ArturFriesen) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift:1772]: My fault, the expired and expireSoon notifications are triggered upfront!

**Review by ArturFriesen — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift (hunk 6 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Developer Menu/Features/DMNotifications/DMNotificationsViewModel.swift:122]: No, its in the dev menu...

File: src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift (hunk 8 above)
Comment (pascal-brause) [reply src/xcode/ENA/ENA/Source/Services/HealthCertificate/__test__/HealthCertificateServiceTests.swift:1772]: Correct.

**Pull request merged**